### PR TITLE
Streamlinbe fetch_attrs interface

### PR DESCRIPTION
**Pull request by simo5**
This allows for writing much shorter functions with a lot less auxiliary variables needed.
It also avoids many manual coding errors by providing more utility functions.

### CODE REVIEW

**Comment by simo5** [comment]
@Jakuje you mentioned on a recent PR that we could improve attribute fetching for keyt objects, as there is a lot of repetition.
This patch goes in that direction.

File: src/exchange.c [hunk 1]
```diff
@@ -291,7 +292,7 @@ static int p11prov_ecdh_derive(void *ctx, unsigned char *secret,
     }
 
     P11PROV_debug("ECDH derived hey handle: %lu", secret_handle);
-    FA_ASSIGN_ALL(attrs[0], CKA_VALUE, &secret, &secret_len, false, true);
+    FA_SET_BUF_VAL(attrs, num, CKA_VALUE, secret, secret_len, false, true);
     ret = p11prov_fetch_attributes(ecdhctx->provctx, session, secret_handle,
                                    attrs, 1);
```
Comment (Jakuje) [inline src/exchange.c:297]:
```suggestion
                                   attrs, num);
```

File: src/kdf.c [hunk 2]
```diff
@@ -154,7 +155,7 @@ static int p11prov_hkdf_derive(void *ctx, unsigned char *key, size_t keylen,
     }
 
     P11PROV_debug("HKDF derived hey handle: %lu", dkey_handle);
-    FA_ASSIGN_ALL(attrs[0], CKA_VALUE, &key, &dkey_len, false, true);
+    FA_SET_BUF_VAL(attrs, num, CKA_VALUE, key, dkey_len, false, true);
     ret = p11prov_fetch_attributes(hkdfctx->provctx, hkdfctx->session,
                                    dkey_handle, attrs, 1);
```
Comment (Jakuje) [inline src/kdf.c:160]:
```suggestion
                                   dkey_handle, attrs, num);
```

**Comment by Jakuje** [comment]
> @Jakuje you mentioned on a recent PR that we could improve attribute fetching for keyt objects, as there is a lot of repetition. This patch goes in that direction.

If you refer to https://github.com/latchset/pkcs11-provider/pull/153#discussion_r1060597349 it was about code repetition when printing ID and label for different key types and not about fetching attrs, but this generally looks ok. But I will have a better look tomorrow.

File: src/objects.c [hunk 3]
```diff
@@ -37,7 +37,7 @@ struct p11prov_obj {
     } data;
 
     CK_ATTRIBUTE *attrs;
-    unsigned long numattrs;
+    int numattrs;
```
Comment (beldmit) [inline src/objects.c:40]: Why this change?

File: src/kdf.c (hunk 2 above)
Comment (simo5) [reply src/kdf.c:160]: Same as above :)

File: src/objects.c (hunk 3 above)
Comment (simo5) [reply src/objects.c:40]: because of p11prov_move_alloc_attrs(), it takes an int, and we really do not need unsigned long many attrs :-)

File: src/exchange.c (hunk 1 above)
Comment (Jakuje) [reply src/exchange.c]: I would rather have larger diff than magic numbers around that you need to change when you add more attrs. But it might be just my preference.

**Review by Jakuje — APPROVED** [review]

**Comment by simo5** [comment]
Thanks Jakub.
